drivers/serial/uart_cdns.c

```diff
@@ -105,21 +105,24 @@ static int uart_cdns_read_fifo(const struct device *dev, uint8_t *rx_data, const
 void uart_cdns_enable_tx_irq(const struct device *dev)
 {
 	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
-
+	unsigned int key;
 	/*
 	 * TX empty interrupt only triggered when TX removes the last byte from the
-	 * TX FIFO. We need another way generate the first interrupt. This is why
-	 * we have the timer involved here
+	 * TX FIFO. We need another way generate the first interrupt. If the TX FIFO
+	 * is already empty, we need to trigger the callback manually.
 	 */
-	uart_regs->rx_timeout = DEFAULT_RTO_PERIODS_FACTOR;
-	uart_regs->intr_enable = CSR_TEMPTY_MASK | CSR_TOUT_MASK;
+	uart_regs->intr_enable = CSR_TEMPTY_MASK;
+	key = irq_lock();
+	uart_cdns_irq_handler(dev);
+	irq_unlock(key);
```

Comment:
Don't you need to check if the FIFO is empty before doing callback? Because the IRQ handler does not check for whether there are any pending interrupts, and simply do the callback.

Reply:
I wasn't sure if that was necessary because for something like the `shell_uart.c`, it should be assumed that it put data in to the ring buffer calling the `uart_irq_tx_enable` and the callback should put more data in to the FIFO of the uart reading out of the ring buffer. If the FIFO wasn't full, when it called the `tx_enable`, then the same thing would happen anyway when it eventually gets empty.

I do see other Uarts have similar approaches...
https://github.com/zephyrproject-rtos/zephyr/blob/193eeaef0c04570b80a954b58ccd98075150fb11/drivers/serial/uart_apbuart.c#L378